hw/top_earlgrey/data/clocks.xdc

```diff
@@ -16,21 +16,31 @@ set clks_aon_unbuf [get_clocks -of_objects [get_pin clkgen/pll/CLKOUT4]]
-create_generated_clock -name clk_io_div4 -source [get_pin ${u_pll}/CLKOUT0] -divide_by 4 [get_pin ${u_div4}/u_clk_div_buf/gen_xilinx.u_impl_xilinx/gen_fpga_buf.bufg_i/O]
+create_generated_clock -name clk_io_div4 -source [get_pin ${u_pll}/CLKOUT0] -divide_by 4 [get_pin ${u_div4}/u_clk_div_buf/gen_xilinx.u_impl_xilinx/gen_fpga_buf.gen_bufg.bufg_i/O]
 
 # the step-down mux is implemented with a LUT right now and the mux switches on the falling edge.
 # therefore, Vivado propagates both clock edges down the clock network.
 # this implementation is not ideal - but we can at least tell Vivado to only honour the rising edge for
 # timing analysis.
-set_clock_sense -positive top_*/u_clkmgr_aon/u_no_scan_io_div2_div/gen_div2.u_div2/gen_fpga_buf.bufg_i_i_1__0/O
+set_clock_sense -positive top_*/u_clkmgr_aon/u_no_scan_io_div2_div/gen_div2.u_div2/gen_fpga_buf.gen_bufg.bufg_i_i_1__0/O
 
-## JTAG and SPI clocks
-create_clock -add -name lc_jtag_tck -period 100.00 -waveform {0 5} [get_pin top_*/u_pinmux_aon/u_pinmux_strap_sampling/u_pinmux_jtag_buf_lc/prim_clock_buf_tck/gen_xilinx.u_impl_xilinx/gen_fpga_buf.bufg_i/O]
-create_clock -add -name rv_jtag_tck -period 100.00 -waveform {0 5} [get_pin top_*/u_pinmux_aon/u_pinmux_strap_sampling/u_pinmux_jtag_buf_rv/prim_clock_buf_tck/gen_xilinx.u_impl_xilinx/gen_fpga_buf.bufg_i/O]
-create_clock -add -name clk_spi_in  -period 100.00 -waveform {0 5} [get_pin top_*/u_spi_device/u_clk_spi_in_buf/gen_xilinx.u_impl_xilinx/gen_fpga_buf.bufg_i/O]
-create_clock -add -name clk_spi_out -period 100.00 -waveform {0 5} [get_pin top_*/u_spi_device/u_clk_spi_out_buf/gen_xilinx.u_impl_xilinx/gen_fpga_buf.bufg_i/O]
+## JTAG clocks
+create_clock -add -name lc_jtag_tck -period 100.00 -waveform {0 5} [get_pin top_*/u_pinmux_aon/u_pinmux_strap_sampling/u_pinmux_jtag_buf_lc/prim_clock_buf_tck/gen_xilinx.u_impl_xilinx/gen_fpga_buf.gen_bufg.bufg_i/O]
+create_clock -add -name rv_jtag_tck -period 100.00 -waveform {0 5} [get_pin top_*/u_pinmux_aon/u_pinmux_strap_sampling/u_pinmux_jtag_buf_rv/prim_clock_buf_tck/gen_xilinx.u_impl_xilinx/gen_fpga_buf.gen_bufg.bufg_i/O]
 
-set_clock_groups -group ${clks_10_unbuf} -group ${clks_48_unbuf} -group ${clks_aon_unbuf} -group clk_io_div2 -group clk_io_div4 -group lc_jtag_tck -group rv_jtag_tck -group clk_spi_in -group clk_spi_out -group sys_clk_pin -asynchronous
+## SPI clocks
+create_clock -add -name clk_spi  -period 100.00 -waveform {0 50} [get_ports SPI_DEV_CLK]
+set_input_delay -clock clk_spi 5 [get_ports SPI_DEV_D0]
+set_input_delay -clock clk_spi 5 [get_ports SPI_DEV_D1]
+set_output_delay -clock clk_spi 5 [get_ports SPI_DEV_D0] -add_delay
+set_output_delay -clock clk_spi 5 [get_ports SPI_DEV_D1] -add_delay
+
+# set clock sense on the input to spi buffers to help the tool understand the clocks are shifted versions of each other
+# This can also be accomplished through create_genearted_clocks.
+set_clock_sense -negative [get_pins top_earlgrey/u_spi_device/gen_fpga_buf.gen_bufr.bufr_i_i_1/O] -clocks clk_spi
```

Comment:
i'm experimenting with create_genearted_clock right now, but somehow the results are worse. I'll need to investigate a bit.

Reply:
Let me know if you'd like another pair of eyes on the constraints :)

Reply:
i did a stupid thing and forgot to include it in the proper clock group..SIGH... the tool was trying to time it with sys_clock or something (no idea, never finished the run). 

The new one looks better. I've switched to `create_generated_clock` also since I think this is cleaner than using `set_clock_sense`. It's also easier to look at hold time paths between the two clocks, which show the half cycle offset.